### ISSUE TITLE
Const/literal type inference

### ISSUE THREAD

**Issue by rChaoz** [microsoft/pyright#8635]
In short, the suggestion is to add something similar to TypeScript's `as const` for type inference, specifically for dicts, as the feature [already exists for tuples](https://microsoft.github.io/pyright/#/type-inference?id=tuple-expressions) and I'm not sure if it makes sense for lists.

**Is your feature request related to a problem? Please describe.**

Would allow to type-check dict spread arguments to functions:
```python
def f(num: int): ...
# No errors here currently
f(**{ "wrong_name": 123 })  # type of dict is inferred to dict[str, int]
```

Would also fix type errors related to this:
```python
def f(num: int, string: str = "default"): ...
# Error: Argument of type "int" cannot be assigned to parameter "string"
f(**{ "num": 123 })  # type of dict is inferred to dict[str, int]
```

Personally I've had this use case when having to programatically instantiate Pydantic models with fields with dash-case names (normally used for parsing YAML files).

**Describe the solution you’d like**

A feature toggle (since otherwise this would be breaking, I believe) to automatically infer the strictest type possible for `Final` variables, or another syntax:

```python
d: Final = { "num": 123, "string": "abc" }
# type should be:
#    dict[Literal["num", "string"], Literal[123, "abc"]]
# even better type but not clue if feasible:
#    TypedDict(num=int, string=str)
```

**Comment by alwaysmpe** [comment]
FYI, a `TypedDict` partially solves this:
```python
from typing import TypedDict, NotRequired

class FArgs(TypedDict):
    num: int
    string: NotRequired[str]

def f(num: int, string: str = "default"): ...
f(**FArgs({ "num": 123 }))  # no type error
f(**FArgs({ "num": "123" }))  # type error
```
alternatively you can use the `inspect` library to build a signature, but I believe this has weaker type checking. I'm not sure if it's possible to extract a typed dictionary from a signature.

**Comment by erictraut** [comment]
There's currently nothing in the [Python typing spec](https://typing.readthedocs.io/en/latest/spec/index.html) that would enable the capability that you're suggesting here. 

`Final` wouldn't be appropriate here because it already has a well-defined meaning that is not applicable here. A variable that has a `Final` qualifier can receive only one assignment, but the object assigned to it is not immutable. In your example, even if the variable `d` is marked `Final`, you can still modify the contents of the dictionary that's assigned to it. For that reason, I don't think it makes sense to ever special-case type inference rules for dict literal expressions even if the value of that expression is assigned to a `Final` type variable.

To do what you're suggesting would require a new feature in the Python type system. If that's something you'd like to explore further, I recommend starting a discussion in the [Python typing forum](https://discuss.python.org/c/typing/32).
